ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/services/commands/tracker/CompletionResponse.scala

```diff
@@ -92,41 +91,44 @@ object CompletionResponse {
   private[daml] def toException(response: TrackedCompletionFailure): StatusException =
     response match {
       case QueueCompletionFailure(failure) =>
-        toException(failure)
+        val metadata = extractMetadata(failure)
+        val statusBuilder = extractStatus(failure)
+        buildException(metadata, statusBuilder)
       case QueueSubmitFailure(status) =>
-        val protoStatus =
-          rpc.Status
-            .newBuilder()
-            .setCode(status.getCode.value())
-            .setMessage(Option(status.getDescription).getOrElse("Failed to submit request"))
-        buildException(Map.empty[String, String], protoStatus)
+        val statusBuilder = GrpcStatus.toJavaBuilder(status)
+        buildException(Map.empty, statusBuilder)
     }
 
-  def toException(response: CompletionResponse.CompletionFailure): StatusException = {
-    val metadata = response.metadata
-    val status = extractStatus(response)
-    buildException(metadata, status)
+  private def extractMetadata(response: CompletionFailure): Map[String, String] = response match {
+    case notOkResponse: CompletionResponse.NotOkResponse => notOkResponse.metadata
+    case _ => Map.empty
+  }
+
+  private def extractStatus(response: CompletionFailure) = response match {
```

Comment:
To improve readability, you might want to add the return type to this signature as 'status' can mean multiple things in this context.